proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,858 @@
+  data IOL a
+  returnL :: a ->. IOL a
+  bindL :: IOL a ->. (a ->. IOL b) ->. IOL b
+
+  -- Definition of sockets
+  data State = Unbound | Bound | Listening | Connected
+  data Socket (s :: State)
+  data SocketAddress
+
+  -- When a (TCP) socket is created it is Unbound.
+  socket :: IOL (Socket Unbound)
+  -- To bind a socket to a port we take an Unbound socket, and make it
+  -- Bound. The type of bindL will ensure that the socket is threaded
+  -- through the computation, so that the (Socket Unbound) is not
+  -- accessible: we cannot bind a socket twice.
+  bind :: Socket Unbound ->. SocketAddress -> IOL (Socket Bound)
+  -- A socket must be bound to a port before we start listening
+  listen :: Socket Bound->. IOL (Socket Listening)
+  -- A socket can accept multiple connection, therefore, the socket is
+  -- returned in the same state by accept. A second, bidirectional,
+  -- socket representing the connection is also returned. Both have to
+  -- be used in a single-threaded fashion.
+  accept :: Socket Listening ->. IOL (Socket Listening, Socket Connected)
+  connect :: Socket Unbound ->. SocketAddress -> IOL (Socket Connected)
+  send :: Socket Connected ->. ByteString -> IOL (Socket Connected, Unrestricted Int)
+  receive :: Socket Connected -> IOL (Socket Connected, Unrestricted ByteString)
+  close :: ∀s. Socket s -> IOL ()
+
+The `paper <https://arxiv.org/abs/1710.09756>`_ mentions other use
+cases as well, such as efficient and safe data serialization as well
+as using types to statically enforce interaction protocols between
+communicating processes (*e.g.* via RPC calls). GHC users will no
+doubt invent many more use cases over time.
+
+Proposed Change Specification
+-----------------------------
+
+We introduce a new language extension. Types with a linearity
+specification are syntactically legal anywhere in a module if and only
+if ``-XLinearTypes`` is turned on.
+
+Definition
+~~~~~~~~~~
+
+We say that a function ``f`` is *linear* when ``f u`` is consumed
+exactly once implies that ``u`` is *consumed exactly once* (defined
+as follows).
+
+- Consuming a value of a data type exactly once means evaluating it to
+  head normal form, then consuming its fields exactly once
+- Consuming a function exactly once means applying it and consuming
+  its result exactly once
+
+The type of linear function from type ``A`` to type ``B`` is written
+``A ->. B`` (see syntax below).
+
+Linearity is a strengthening of the contract of the regular function
+type ``A -> B``, which will be called the type of *unrestricted*
+functions.
+
+Remark: linear function ``f`` can diverge or be called on diverging
+data. It may feel weird because ``f`` will not necessarily consume its
+argument. But it's alright: the metatheory still works. One can think
+of a diverging computation as not consuming its result exactly once,
+therefore, it is vacuously true that if ``f u`` is consumed exactly
+once, then ``u`` is consumed exactly once.
+
```

Comment:
That said, in practice, the type-system will work as one can reasonably expect even when a function diverges. For example, if a linear function is fed an infinite list, the list won't be traversed more than once. Another example: even a diverging linear function cannot pattern match several times on the same portion of its input.